### PR TITLE
Treat empty cells in CSV as default values.

### PR DESCRIPTION
**Pull request by akuzm**
https://github.com/yandex/ClickHouse/issues/5349

For changelog. Remove if this is non-significant change.
```
When inserting data in CSV format, empty unquoted input values are replaced with default values for the respective columns, if [input_format_defaults_for_omitted_fields](../operations/settings/settings.md#session_settings-input_format_defaults_for_omitted_fields) is enabled.

При импорте данных в формате CSV пустые ячейки без кавычек заменяются значениями по умолчанию для соответствующих колонок, если включён параметр [input_format_defaults_for_omitted_fields](../operations/settings/settings.md#session_settings-input_format_defaults_for_omitted_fields).
```

Category (leave one):
- Improvement

### CODE REVIEW

File: dbms/src/Formats/CSVRowInputStream.cpp [hunk 1]
```diff
@@ -191,46 +162,104 @@ void CSVRowInputStream::readPrefix()
...
+            if (!read_columns[i])
+            {
+                ///!!!FIXME
+                /// 1. Why are the defaults inserted by every row source,
+                /// and not by the caller, given that it walks this array
+                /// anyway?
+                /// 2. The value is overwritten by
+                /// AddingDefaultsBlockInputStream, so we don't care what it
+                /// is. Should we use IColumn::insertDefault() instead?
+                data_types[i]->insertDefaultInto(*columns[i]);
```
Comment (akuzm) [inline dbms/src/Formats/CSVRowInputStream.cpp]: Does anybody know the answer to the above questions?

File: docs/en/operations/settings/settings.md [hunk 2]
```diff
@@ -186,20 +186,15 @@ Ok.
...
-
-When executing the `INSERT` query, the ClickHouse client prepares data and sends it to the server for writing. The client gets the table structure from the server when preparing the data. In some cases, the client needs more information than the server sends by default. Turn on the extended data exchange with `input_format_defaults_for_omitted_fields = 1`.
-
-When the extended data exchange is enabled, the server sends the additional metadata along with the table structure. The composition of the metadata depends on the operation.
-
-Operations where you may need the extended data exchange enabled:
-
-- Inserting data in [JSONEachRow](../../interfaces/formats.md#jsoneachrow) format.
-
-For all other operations, ClickHouse doesn't apply the setting.
```
Comment (akuzm) [inline docs/en/operations/settings/settings.md]: I deleted this big explanation about the extended data exchange and slightly extended the note about performance below. This option replaces empty cells with defaults, and the detailed description of how it's implemented doesn't make any difference to the user.

File: dbms/src/Formats/CSVRowInputStream.cpp [hunk 3]
```diff
@@ -191,46 +162,104 @@ void CSVRowInputStream::readPrefix()
...
     {
         const auto & column_index = column_indexes_for_input_fields[input_position];
         if (column_index)
         {
-            skipWhitespacesAndTabs(istr);
-            data_types[*column_index]->deserializeAsTextCSV(*columns[*column_index], istr, format_settings);
-            skipWhitespacesAndTabs(istr);
+            const auto & type = data_types[*column_index];
+            if (*istr.position() == delimiter
+                    && format_settings.csv.empty_as_default)
```
Comment (alexey-milovidov) [inline dbms/src/Formats/CSVRowInputStream.cpp]: Do we check for empty values at the last column?

```
INSERT INTO table (a, b) FORMAT CSV
Hello,
```

Comment (alexey-milovidov) [reply dbms/src/Formats/CSVRowInputStream.cpp]: (it requires comparison with \r and \n; we assume that unquoted newlines are not possible in value)

File: dbms/src/Formats/CSVRowInputStream.cpp (hunk 1 above)
Comment (alexey-milovidov) [reply dbms/src/Formats/CSVRowInputStream.cpp]: If I remebmer correctly:

We need to return block with columns of corresponding size. Non-existing values should be prefilled nevertheless. Then we fill default values through applying `if` function or in similar way (that expect block with proper column sizes).

Comment (alexey-milovidov) [reply dbms/src/Formats/CSVRowInputStream.cpp]: (you can update your comment to my)

**Pull request merged**